firefox-ios/Storage/Rust/RustRemoteTabs.swift

```diff
@@ -165,6 +172,207 @@ public class RustRemoteTabs {
+                    } else {
+                        let errHeader = "Failed to set some pending commands as sent:\n"
+                        let errMessages: String = errors.reduce(errHeader, { result, err in
+                            result.appending("  - \(err.localizedDescription)\n")
+                        })
+                        self.logger.log(errMessages,
+                                        level: .warning,
+                                        category: .tabs)
+                    }
+                }
+            }
+        }
+    }
+
+    private func getUnsentCommandsByDeviceId(deviceId: String, completion: @escaping ([PendingCommand]) -> Void) {
+        guard let commandQueue = self.commandQueue else {
+            let err = TabsApiError.UnexpectedTabsError(reason: "Command queue is not initialized") as MaybeErrorType
+            self.logger.log(err.description,
+                            level: .warning,
+                            category: .tabs)
+            completion([PendingCommand]())
+            return
+        }
+
+        commandQueue.getUnsentCommands { result in
+            switch result {
+            case .success(let commands):
+                let filteredCommands = commands.filter { $0.deviceId == deviceId }
+                completion(filteredCommands)
+            case .failure(let error):
+                self.logger.log("Failed to get unsent commands: \(String(describing: error))",
+                                level: .warning,
+                                category: .tabs)
+                completion([PendingCommand]())
+            }
+        }
+    }
+
+    private func handleCommandQueueErrors(result: Result<Bool, any Error>) {
+        switch result {
+        case .success(let didSucceed):
+            if !didSucceed {
+                self.logger.log("Unexpected error while attempting to update command queue",
+                                level: .warning,
+                                category: .tabs)
+            }
+        case .failure(let error):
+            self.logger.log("Failed to update command queue: \(String(describing: error))",
+                            level: .warning,
+                            category: .tabs)
+        }
+    }
+}
+
+internal class RemoteTabsCommandQueue {
+    var commandStore: RemoteCommandStore?
+
+    init() {}
+
+    func openCommandStore(tabsStore: TabsStore) {
+        self.commandStore = tabsStore.newRemoteCommandStore()
+    }
+
+    func addRemoteCommand(deviceId: String,
+                          command: RemoteCommand,
+                          completion: @escaping (Result<Bool, Error>) -> Void) {
+        guard let commandStore = self.commandStore else {
+            completion(.failure(TabsApiError.UnexpectedTabsError(reason: "Command store is not initialized")))
+            return
+        }
+
+        do {
+            let didSucceed = try commandStore.addRemoteCommand(deviceId: deviceId, command: command)
```

Comment:
Ahh, `false` doesn't mean that it failed to add the command; it means the command was already in the queue. It looks like we only issue a log if `addRemoteCommand` returns `false`, though; and nothing else?

Reply:
It logs if the rust component `addRemoteCommand` function returns `false` or an error via `handleCommandQueueErrors`. However, I'll refactor this so it's clearer. Also the remove the log for language for the `false` case.

Reply:
Oh . . . no I see it's the `.success` vs `.failure`.

Reply:
https://github.com/mozilla-mobile/firefox-ios/pull/22799/commits/7d44018db816a8b4494ca1d7b1f12e8113690ce9